File: PaperPiano/js/audio-engine.js
/**
 * AudioEngine - Web Audio API synthesis for musical notes.
 * Uses multiple oscillators with ADSR envelope for rich, low-latency sound.
 */

const NOTE_FREQUENCIES = {
    'C': 261.63, 'C#': 277.18, 'Db': 277.18,
    'D': 293.66, 'D#': 311.13, 'Eb': 311.13,
    'E': 329.63,
    'F': 349.23, 'F#': 369.99, 'Gb': 369.99,
    'G': 392.00, 'G#': 415.30, 'Ab': 415.30,
    'A': 440.00, 'A#': 466.16, 'Bb': 466.16,
    'B': 493.88
};

// Waveform presets for different instruments
const PRESETS = {
    piano: {
        oscillators: [
            { type: 'triangle', detune: 0, gain: 0.35 },
            { type: 'sine', freqMul: 2, gain: 0.08 },
            { type: 'sine', freqMul: 0.5, gain: 0.04 }
        ],
        attack: 0.008,
        decay: 0.25,
        sustain: 0.15,
        release: 0.2
    },
    organ: {
        oscillators: [
            { type: 'sine', detune: 0, gain: 0.25 },
            { type: 'sine', freqMul: 2, gain: 0.15 },
            { type: 'sine', freqMul: 3, gain: 0.08 },
            { type: 'sine', freqMul: 4, gain: 0.04 }
        ],
        attack: 0.05,
        decay: 0.1,
        sustain: 0.3,
        release: 0.1
    },
    synth: {
        oscillators: [
            { type: 'sawtooth', detune: 0, gain: 0.2 },
            { type: 'sawtooth', detune: 7, gain: 0.15 },
            { type: 'square', freqMul: 0.5, gain: 0.05 }
        ],
        attack: 0.005,
        decay: 0.15,
        sustain: 0.25,
        release: 0.3
    }
};

export class AudioEngine {
    constructor() {
        this.ctx = null;
        this.masterGain = null;
        this.compressor = null;
        this.activeNotes = new Map();
        this.preset = 'piano';
    }

    init() {
        if (this.ctx) return;
        this.ctx = new (window.AudioContext || window.webkitAudioContext)({
            latencyHint: 'interactive',
            sampleRate: 44100
        });

        // Dynamics compressor prevents clipping with many simultaneous notes
        this.compressor = this.ctx.createDynamicsCompressor();
        this.compressor.threshold.value = -20;
        this.compressor.knee.value = 20;
        this.compressor.ratio.value = 8;
        this.compressor.attack.value = 0.002;
        this.compressor.release.value = 0.15;

        this.masterGain = this.ctx.createGain();
        this.masterGain.gain.value = 0.6;

        this.compressor.connect(this.masterGain);
        this.masterGain.connect(this.ctx.destination);
    }

    resume() {
        if (this.ctx?.state === 'suspended') {
            this.ctx.resume();
        }
    }

    setPreset(name) {
        if (PRESETS[name]) this.preset = name;
    }

    /**
     * Start playing a note for a given shape.
     * If the shape is already playing, this is a no-op (prevents re-triggering).
     */
    noteOn(shapeId, noteName) {
        if (!this.ctx) this.init();
        this.resume();

        if (this.activeNotes.has(shapeId)) return;

        const freq = this.getFrequency(noteName);
        if (!freq) return;

        const preset = PRESETS[this.preset];
        const now = this.ctx.currentTime;
        const nodes = [];

        for (const oscDef of preset.oscillators) {
            const osc = this.ctx.createOscillator();
            osc.type = oscDef.type;

            const oscFreq = oscDef.freqMul ? freq * oscDef.freqMul : freq;
            osc.frequency.setValueAtTime(oscFreq, now);
            if (oscDef.detune) osc.detune.setValueAtTime(oscDef.detune, now);

            const gain = this.ctx.createGain();
            gain.gain.setValueAtTime(0, now);
            // Attack
            gain.gain.linearRampToValueAtTime(oscDef.gain, now + preset.attack);
            // Decay to sustain
            gain.gain.exponentialRampToValueAtTime(
                Math.max(oscDef.gain * (preset.sustain / 0.35), 0.001),
                now + preset.attack + preset.decay
            );

            osc.connect(gain);
            gain.connect(this.compressor);
            osc.start(now);

            nodes.push({ osc, gain });
        }

        this.activeNotes.set(shapeId, { nodes, preset });
    }

    /**
     * Stop a note for a given shape with a smooth release.
     */
    noteOff(shapeId) {
        const entry = this.activeNotes.get(shapeId);
        if (!entry) return;

        const now = this.ctx.currentTime;
        const release = entry.preset.release;

        for (const { osc, gain } of entry.nodes) {
            gain.gain.cancelScheduledValues(now);
            gain.gain.setValueAtTime(gain.gain.value, now);
            gain.gain.linearRampToValueAtTime(0.001, now + release);
            osc.stop(now + release + 0.02);
        }

        this.activeNotes.delete(shapeId);
    }

    noteOffAll() {
        for (const id of [...this.activeNotes.keys()]) {
            this.noteOff(id);
        }
    }

    getFrequency(noteName) {
        if (!noteName) return null;
        const name = noteName.trim();

        if (NOTE_FREQUENCIES[name]) return NOTE_FREQUENCIES[name];

        // Try uppercase
        const upper = name.toUpperCase();
        for (const [key, val] of Object.entries(NOTE_FREQUENCIES)) {
            if (key.toUpperCase() === upper) return val;
        }

        // Try just the first letter
        const letter = upper.charAt(0);
        if (NOTE_FREQUENCIES[letter]) return NOTE_FREQUENCIES[letter];

        return null;
    }
}

File: PaperPiano/audio_engine.py
"""
AudioEngine — Pure-Python tone synthesis using the built-in audioop / wave
approach piped through PyAudio *or* a cross-platform fallback via
winsound (Windows) or simpleaudio.

Uses numpy to generate ADSR-enveloped waveforms in real time.
Each note-on spawns a background thread that writes samples
continuously until note-off requests a release.
"""

import math
import threading
import struct
import time
import numpy as np

# ── Note frequencies (C4-B4) ──────────────────────────────────────────────────
NOTE_FREQUENCIES = {
    'C': 261.63, 'C#': 277.18, 'Db': 277.18,
    'D': 293.66, 'D#': 311.13, 'Eb': 311.13,
    'E': 329.63,
    'F': 349.23, 'F#': 369.99, 'Gb': 369.99,
    'G': 392.00, 'G#': 415.30, 'Ab': 415.30,
    'A': 440.00, 'A#': 466.16, 'Bb': 466.16,
    'B': 493.88,
}

# ── Waveform helpers ──────────────────────────────────────────────────────────
def _sine(t, freq):
    return np.sin(2 * np.pi * freq * t)

def _triangle(t, freq):
    return 2 * np.abs(2 * (t * freq - np.floor(t * freq + 0.5))) - 1

def _sawtooth(t, freq):
    return 2 * (t * freq - np.floor(t * freq + 0.5))

def _square(t, freq):
    return np.sign(np.sin(2 * np.pi * freq * t))

WAVE_FN = {
    'sine': _sine,
    'triangle': _triangle,
    'sawtooth': _sawtooth,
    'square': _square,
}

# ── Presets (mirror the JS version) ──────────────────────────────────────────
PRESETS = {
    'piano': {
        'oscillators': [
            {'type': 'triangle', 'detune': 0, 'gain': 0.35},
            {'type': 'sine', 'freq_mul': 2, 'gain': 0.08},
            {'type': 'sine', 'freq_mul': 0.5, 'gain': 0.04},
        ],
        'attack': 0.008,
        'decay': 0.25,
        'sustain': 0.15,
        'release': 0.2,
    },
    'organ': {
        'oscillators': [
            {'type': 'sine', 'detune': 0, 'gain': 0.25},
            {'type': 'sine', 'freq_mul': 2, 'gain': 0.15},
            {'type': 'sine', 'freq_mul': 3, 'gain': 0.08},
            {'type': 'sine', 'freq_mul': 4, 'gain': 0.04},
        ],
        'attack': 0.05,
        'decay': 0.1,
        'sustain': 0.3,
        'release': 0.1,
    },
    'synth': {
        'oscillators': [
            {'type': 'sawtooth', 'detune': 0, 'gain': 0.2},
            {'type': 'sawtooth', 'detune': 7, 'gain': 0.15},
            {'type': 'square', 'freq_mul': 0.5, 'gain': 0.05},
        ],
        'attack': 0.005,
        'decay': 0.15,
        'sustain': 0.25,
        'release': 0.3,
    },
}

SAMPLE_RATE = 44100
CHUNK = 1024

# ── Try to import a suitable audio backend ───────────────────────────────────
_backend = None
try:
    import pyaudio                       # pip install pyaudio
    _backend = 'pyaudio'
except ImportError:
    pass

if _backend is None:
    try:
        import simpleaudio               # pip install simpleaudio
        _backend = 'simpleaudio'
    except ImportError:
        pass


class AudioEngine:
    """Low-latency multi-note synthesiser with ADSR envelope."""

    def __init__(self):
        self.preset_name = 'piano'
        self._active: dict[str, dict] = {}   # shape_id → control dict
        self._lock = threading.Lock()
        self._stream = None
        self._pa = None
        self._running = False
        self._audio_thread = None

    # ── lifecycle ─────────────────────────────────────────────────────────────
    def init(self):
        if self._running:
            return
        self._running = True
        if _backend == 'pyaudio':
            self._pa = pyaudio.PyAudio()
            self._stream = self._pa.open(
                format=pyaudio.paFloat32,
                channels=1,
                rate=SAMPLE_RATE,
                output=True,
                frames_per_buffer=CHUNK,
                stream_callback=self._pyaudio_cb,
            )
            self._stream.start_stream()
        else:
            # Real-time fallback: generate chunks in a thread
            self._audio_thread = threading.Thread(target=self._audio_loop, daemon=True)
            self._audio_thread.start()

    def _pyaudio_cb(self, in_data, frame_count, time_info, status):
        import pyaudio as _pa
        data = self._render(frame_count)
        return (data.tobytes(), _pa.paContinue)

    def _audio_loop(self):
        """Fallback loop for simpleaudio / winsound — not ideal latency."""
        if _backend == 'simpleaudio':
            while self._running:
                with self._lock:
                    if not self._active:
                        time.sleep(0.01)
                        continue
                samples = self._render(CHUNK * 4)
                pcm = (samples * 32767).astype(np.int16)
                try:
                    play = simpleaudio.play_buffer(pcm, 1, 2, SAMPLE_RATE)
                    play.wait_done()
                except Exception:
                    time.sleep(0.01)

    def _render(self, n_frames: int) -> np.ndarray:
        """Mix all active notes into a single float32 buffer."""
        buf = np.zeros(n_frames, dtype=np.float32)
        t = np.arange(n_frames, dtype=np.float64) / SAMPLE_RATE
        dead = []
        with self._lock:
            for sid, ctrl in self._active.items():
                phase = ctrl['phase']
                freq = ctrl['freq']
                preset = ctrl['preset']
                releasing = ctrl['releasing']

                osc_sum = np.zeros(n_frames, dtype=np.float64)
                for odef in preset['oscillators']:
                    fn = WAVE_FN.get(odef['type'], _sine)
                    f = odef.get('freq_mul', 1.0) * freq
                    detune_cents = odef.get('detune', 0)
                    if detune_cents:
                        f *= 2 ** (detune_cents / 1200)
                    osc_sum += fn(t + phase, f) * odef['gain']

                # Simple ADSR envelope
                env = self._envelope(ctrl, n_frames, preset)
                osc_sum *= env

                buf += osc_sum.astype(np.float32)
                ctrl['phase'] += n_frames / SAMPLE_RATE
                ctrl['env_pos'] += n_frames

                if releasing and ctrl.get('env_done', False):
                    dead.append(sid)

            for sid in dead:
                self._active.pop(sid, None)

        # Soft-clip
        buf = np.tanh(buf * 0.8) * 0.6
        return buf

    def _envelope(self, ctrl, n_frames, preset):
        """Per-sample ADSR envelope."""
        a = int(preset['attack'] * SAMPLE_RATE)
        d = int(preset['decay'] * SAMPLE_RATE)
        s = preset['sustain']
        r = int(preset['release'] * SAMPLE_RATE)

        env = np.ones(n_frames, dtype=np.float64)
        pos = ctrl['env_pos']

        if ctrl['releasing']:
            rel_pos = ctrl.get('rel_pos', 0)
            for i in range(n_frames):
                p = rel_pos + i
                if p < r:
                    env[i] = ctrl['rel_level'] * (1.0 - p / r)
                else:
                    env[i] = 0.0
                    ctrl['env_done'] = True
            ctrl['rel_pos'] = rel_pos + n_frames
        else:
            for i in range(n_frames):
                p = pos + i
                if p < a:
                    env[i] = p / a if a > 0 else 1.0
                elif p < a + d:
                    frac = (p - a) / d if d > 0 else 1.0
                    env[i] = 1.0 - frac * (1.0 - s)
                else:
                    env[i] = s

        return env

    # ── public API ────────────────────────────────────────────────────────────
    def set_preset(self, name: str):
        if name in PRESETS:
            self.preset_name = name

    def note_on(self, shape_id: str, note_name: str):
        if not self._running:
            self.init()
        freq = self.get_frequency(note_name)
        if freq is None:
            return
        with self._lock:
            if shape_id in self._active:
                return  # already playing
            self._active[shape_id] = {
                'freq': freq,
                'preset': PRESETS[self.preset_name],
                'phase': 0.0,
                'env_pos': 0,
                'releasing': False,
                'rel_level': PRESETS[self.preset_name]['sustain'],
                'rel_pos': 0,
                'env_done': False,
            }

    def note_off(self, shape_id: str):
        with self._lock:
            ctrl = self._active.get(shape_id)
            if ctrl and not ctrl['releasing']:
                ctrl['releasing'] = True
                ctrl['rel_pos'] = 0
                # Capture current envelope level for smooth release
                preset = ctrl['preset']
                a = int(preset['attack'] * SAMPLE_RATE)
                d = int(preset['decay'] * SAMPLE_RATE)
                s = preset['sustain']
                pos = ctrl['env_pos']
                if pos < a:
                    ctrl['rel_level'] = pos / a if a > 0 else 1.0
                elif pos < a + d:
                    frac = (pos - a) / d if d > 0 else 1.0
                    ctrl['rel_level'] = 1.0 - frac * (1.0 - s)
                else:
                    ctrl['rel_level'] = s

    def note_off_all(self):
        with self._lock:
            for sid in list(self._active.keys()):
                ctrl = self._active[sid]
                if not ctrl['releasing']:
                    ctrl['releasing'] = True
                    ctrl['rel_pos'] = 0
                    ctrl['rel_level'] = PRESETS[self.preset_name]['sustain']

    def get_frequency(self, note_name: str):
        if not note_name:
            return None
        name = note_name.strip()
        if name in NOTE_FREQUENCIES:
            return NOTE_FREQUENCIES[name]
        upper = name.upper()
        for k, v in NOTE_FREQUENCIES.items():
            if k.upper() == upper:
                return v
        letter = upper[0] if upper else ''
        return NOTE_FREQUENCIES.get(letter)

    def destroy(self):
        self._running = False
        if self._stream:
            self._stream.stop_stream()
            self._stream.close()
        if self._pa:
            self._pa.terminate()
